legacy/builder/constants/constants.go

```diff
@@ -106,10 +104,10 @@ var MSG_LIB_LEGACY = tr("(legacy)")
 var MSG_LIBRARIES_MULTIPLE_LIBS_FOUND_FOR = tr("Multiple libraries were found for \"{0}\"")
 var MSG_LIBRARIES_NOT_USED = tr(" Not used: {0}")
 var MSG_LIBRARIES_USED = tr(" Used: {0}")
-var MSG_LIBRARY_CAN_USE_SRC_AND_UTILITY_FOLDERS = fmt.Sprintf(tr("Library can't use both '%[1]s' and '%[2]s' folders. Double check {0}"), "src", "utility")
+var MSG_LIBRARY_CAN_USE_SRC_AND_UTILITY_FOLDERS = tr("Library can't use both '%[1]s' and '%[2]s' folders. Double check {0}", "src", "utility")
```

Comment:
Here we have these "mixed" messages with the gettext-style `{0}` markers and the printf-style `%[1]s` markers.

AFAIR this was done at the times to let the Java IDE translate the messages.
Basically in the `arduino-builder`, using the legacy `MachineLogger`, the messages are passed as-is (i.e. with the markers `{0}` left intact) together with the value to substitute. It's the duty of the Java IDE to translate the message and replace the markers with the given values.

```
$ arduino-builder -compile -logger=machine -hardware ... -tools ... -libraries ... -fqbn=arduino:avr:uno -ide-version=10815 -build-path /tmp/arduino_build_474903 -verbose /tmp/sketch_sep02a/sketch_sep02a.ino
[.....]
===info ||| Progress {0} ||| [25.00]
===info ||| Progress {0} ||| [50.00]
===info ||| Progress {0} ||| [75.00]
===info ||| Sketch uses {0} bytes ({2}%%) of program storage space. Maximum is {1} bytes. ||| [444 32256 1]
===info ||| Global variables use {0} bytes ({2}%%) of dynamic memory, leaving {3} bytes for local variables. Maximum is {1} bytes. ||| [9 2048 0 2039]
===info ||| Progress {0} ||| [100.00]
```

This must be completely replaced and the translation should be made in the CLI.
I see also that if I use the `HumanLogger` with the parameter `-logger=human` then the output is:
```
$ arduino-builder -compile -logger=human -hardware ... -tools ... -libraries ... -fqbn=arduino:avr:uno -ide-version=10815 -build-path /tmp/arduino_build_474903 -verbose /tmp/sketch_sep02a/sketch_sep02a.ino
Sketch uses 444 bytes (1%) of program storage space. Maximum is 32256 bytes.
Global variables use 9 bytes (0%) of dynamic memory, leaving 2039 bytes for local variables. Maximum is 2048 bytes.
```

so the HumanLogger is able to replace the `{0}`, at this point I'm wondering if we should keep the `%s` or use the gettext notation.

Reply:
Yeah, these are very ugly. I remembered this comment, so didn't worry about them too much because I figured they were going to go away at some point:
https://github.com/arduino/arduino-cli/pull/961#discussion_r512819675

> I'm wondering if we should keep the %s or use the gettext notation.

I think people contributing to the code will be more likely to be familiar with the printf style. I don't know which would be friendly to the translators though. Something we noticed is that Transifex recognizes some of these printf verbs and makes them into a token that is treated like a single character in the editor. But it doesn't recognize all of them. For example, here you can see it gave the `%s` special treatment, but not the `%v`:
![Clipboard01](https://user-images.githubusercontent.com/8572152/131837002-903ca6c4-7103-40f6-813a-30bffed23894.png)
But I see that it should be possible to configure Transifex to recognize whatever syntax we use for placeholders:
https://docs.transifex.com/translation-checks/setting-translation-checks
So this means that it's probably not so important that the translator be familiar with the exact placeholder syntax in use, just as long as they can recognize the significance of the index numbers in the placeholders.

---
FYI, these particular strings are not correctly handled by the `i18n` package because they are declared at package level before the call to `i18n.Init()`, and so are just returned as-is. Details [here](https://arduino.atlassian.net/browse/ATL-1601).

Reply:
Ok it seems that I can't put the panic in the Tr function otherwise the unit-tests will fail:
```
=== RUN   TestBoardOptions
--- FAIL: TestBoardOptions (0.00s)
panic: i18n.Tr called before i18n.Init() [recovered]
	panic: i18n.Tr called before i18n.Init()

goroutine 9 [running]:
testing.tRunner.func1.2(0xc56e00, 0xeaee20)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc0002a2f00)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1146 +0x4b6
panic(0xc56e00, 0xeaee20)
	/opt/hostedtoolcache/go/1.16.7/x64/src/runtime/panic.go:965 +0x1b9
github.com/arduino/arduino-cli/i18n.Tr(0xdceacd, 0x28, 0x0, 0x0, 0x0, 0x0, 0x1)
	/home/runner/work/arduino-cli/arduino-cli/i18n/i18n.go:54 +0xd9
github.com/arduino/arduino-cli/arduino/cores.(*Board).GetBuildProperties(0x13bf060, 0xc0003b6150, 0xc000339400, 0x0, 0x0)
	/home/runner/work/arduino-cli/arduino-cli/arduino/cores/board.go:109 +0x69d
github.com/arduino/arduino-cli/arduino/cores.(*Board).GeneratePropertiesForConfiguration(0x13bf060, 0xdb5f54, 0xe, 0xc000309ef0, 0xc911e0, 0xc0003b6000)
	/home/runner/work/arduino-cli/arduino-cli/arduino/cores/board.go:141 +0x28f
github.com/arduino/arduino-cli/arduino/cores.TestBoardOptions(0xc0002a2f00)
	/home/runner/work/arduino-cli/arduino-cli/arduino/cores/board_test.go:298 +0x4bd
testing.tRunner(0xc0002a2f00, 0xe052a8)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1238 +0x2b3
FAIL	github.com/arduino/arduino-cli/arduino/cores	0.021s
=== RUN   TestIndexParsing
--- FAIL: TestIndexParsing (0.00s)
panic: i18n.Tr called before i18n.Init() [recovered]
	panic: i18n.Tr called before i18n.Init()
```

so I'll remove it for now, we need to come up with a better idea.
BTW it has been very helpful to find all occurrences of the bug, very neat!